File: tests/output/dart/test/client/search_test.dart
import 'package:algolia_client_search/algolia_client_search.dart';
import 'package:algolia_test/algolia_test.dart';
import 'package:test/test.dart';
import 'package:test_api/hooks.dart';

void main() {
  test('calls api with correct read host', () {
    final client = SearchClient(
      appId: "test-app-id",
      apiKey: "test-api-key",
    );
    runTest(
      builder: (requester) => SearchClient(
        appId: client.appId,
        apiKey: client.apiKey,
        options: ClientOptions(requester: requester),
      ),
      call: (client) => client.get(
        path: "/test",
      ),
      intercept: (request) {
        expect(request.host.url, 'test-app-id-dsn.algolia.net');
      },
    );
  });
  test('calls api with correct write host', () {
    final client = SearchClient(
      appId: "test-app-id",
      apiKey: "test-api-key",
    );
    runTest(
      builder: (requester) => SearchClient(
        appId: client.appId,
        apiKey: client.apiKey,
        options: ClientOptions(requester: requester),
      ),
      call: (client) => client.post(
        path: "/test",
      ),
      intercept: (request) {
        expect(request.host.url, 'test-app-id.algolia.net');
      },
    );
  });

  test('calls api with correct user agent', () {
    final client = SearchClient(
      appId: 'appId',
      apiKey: 'apiKey',
    );
    runTest(
      builder: (requester) => SearchClient(
        appId: client.appId,
        apiKey: client.apiKey,
        options: ClientOptions(requester: requester),
      ),
      call: (client) => client.post(
        path: "/test",
      ),
      intercept: (request) {
        TestHandle.current.markSkipped('User agent added using an interceptor');
      },
    );
  });
  test('calls api with default read timeouts', () {
    final client = SearchClient(
      appId: 'appId',
      apiKey: 'apiKey',
    );
    runTest(
      builder: (requester) => SearchClient(
        appId: client.appId,
        apiKey: client.apiKey,
        options: ClientOptions(requester: requester),
      ),
      call: (client) => client.get(
        path: "/test",
      ),
      intercept: (request) {
        expect(5000, request.timeout.inMilliseconds);
      },
    );
  });
  test('calls api with default write timeouts', () {
    final client = SearchClient(
      appId: 'appId',
      apiKey: 'apiKey',
    );
    runTest(
      builder: (requester) => SearchClient(
        appId: client.appId,
        apiKey: client.apiKey,
        options: ClientOptions(requester: requester),
      ),
      call: (client) => client.post(
        path: "/test",
      ),
      intercept: (request) {
        expect(30000, request.timeout.inMilliseconds);
      },
    );
  });

  test('client throws with invalid parameters', () {
    expectError(
      '`appId` is missing.',
      () {
        final client = SearchClient(
          appId: "",
          apiKey: "",
        );
      },
    );
    expectError(
      '`appId` is missing.',
      () {
        final client = SearchClient(
          appId: "",
          apiKey: "my-api-key",
        );
      },
    );
    expectError(
      '`apiKey` is missing.',
      () {
        final client = SearchClient(
          appId: "my-app-id",
          apiKey: "",
        );
      },
    );
  });
  test('`addApiKey` throws with invalid parameters', () {
    final client = SearchClient(
      appId: 'appId',
      apiKey: 'apiKey',
    );
    expectError(
      'Parameter `apiKey` is required when calling `addApiKey`.',
      () {
        return client.addApiKey(
          apiKey: empty(),
        );
      },
    );
  });
  test('`addOrUpdateObject` throws with invalid parameters', () {
    final client = SearchClient(
      appId: 'appId',
      apiKey: 'apiKey',
    );
    expectError(
      'Parameter `indexName` is required when calling `addOrUpdateObject`.',
      () {
        return client.addOrUpdateObject(
          indexName: empty(),
          objectID: "my-object-id",
          body: {},
        );
      },
    );
    expectError(
      'Parameter `objectID` is required when calling `addOrUpdateObject`.',
      () {
        return client.addOrUpdateObject(
          indexName: "my-index-name",
          objectID: empty(),
          body: {},
        );
      },
    );
    expectError(
      'Parameter `body` is required when calling `addOrUpdateObject`.',
      () {
        return client.addOrUpdateObject(
          indexName: "my-index-name",
          objectID: "my-object-id",
          body: empty(),
        );
      },
    );
  });
}

File: tests/output/kotlin/src/commonTest/kotlin/com/algolia/client/SearchTest.kt
package com.algolia.client

import com.algolia.client.api.SearchClient
import com.algolia.client.configuration.*
import com.algolia.client.model.search.*
import com.algolia.client.transport.*
import com.algolia.utils.*
import io.ktor.http.*
import kotlinx.coroutines.test.*
import kotlinx.serialization.json.*
import kotlin.test.*

class SearchTest {

  @Test
  fun `calls api with correct read host`() = runTest {
    val client = SearchClient(appId = "test-app-id", apiKey = "test-api-key")
    client.runTest(
      call = {
        get(
          path = "/test",
        )
      },
      intercept = {
        assertEquals("test-app-id-dsn.algolia.net", it.url.host)
      },
    )
  }

  @Test
  fun `calls api with correct write host`() = runTest {
    val client = SearchClient(appId = "test-app-id", apiKey = "test-api-key")
    client.runTest(
      call = {
        post(
          path = "/test",
        )
      },
      intercept = {
        assertEquals("test-app-id.algolia.net", it.url.host)
      },
    )
  }

  @Test
  fun `calls api with correct user agent`() = runTest {
    val client = SearchClient(appId = "appId", apiKey = "apiKey")
    client.runTest(
      call = {
        post(
          path = "/test",
        )
      },
      intercept = {
        val regexp = "^Algolia for Kotlin \\(\\d+\\.\\d+\\.\\d+(-.*)?\\)(; [a-zA-Z. ]+ (\\(\\d+((\\.\\d+)?\\.\\d+)?(-.*)?\\))?)*(; Search (\\(\\d+\\.\\d+\\.\\d+(-.*)?\\)))(; [a-zA-Z. ]+ (\\(\\d+((\\.\\d+)?\\.\\d+)?(-.*)?\\))?)*$".toRegex()
        val header = it.headers["User-Agent"].orEmpty()
        assertTrue(actual = header.matches(regexp), message = "Expected $header to match the following regex: $regexp")
      },
    )
  }

  @Test
  fun `calls api with default read timeouts`() = runTest {
    val client = SearchClient(appId = "appId", apiKey = "apiKey")
    client.runTest(
      call = {
        get(
          path = "/test",
        )
      },
      intercept = {
        assertEquals(2000, it.connectTimeout)
        assertEquals(5000, it.socketTimeout)
      },
    )
  }

  @Test
  fun `calls api with default write timeouts`() = runTest {
    val client = SearchClient(appId = "appId", apiKey = "apiKey")
    client.runTest(
      call = {
        post(
          path = "/test",
        )
      },
      intercept = {
        assertEquals(2000, it.connectTimeout)
        assertEquals(30000, it.socketTimeout)
      },
    )
  }

  @Test
  fun `client throws with invalid parameters`() = runTest {
    assertFails {
      val client = SearchClient(appId = "", apiKey = "")
    }.let { error -> assertError(error, "`appId` is missing.") }
    assertFails {
      val client = SearchClient(appId = "", apiKey = "my-api-key")
    }.let { error -> assertError(error, "`appId` is missing.") }
    assertFails {
      val client = SearchClient(appId = "my-app-id", apiKey = "")
    }.let { error -> assertError(error, "`apiKey` is missing.") }
  }

  @Test
  fun `'addApiKey' throws with invalid parameters`() = runTest {
    val client = SearchClient(appId = "appId", apiKey = "apiKey")
    assertFails {
      client.addApiKey(
        apiKey = empty(),
      )
    }.let { error -> assertError(error, "Parameter `apiKey` is required when calling `addApiKey`.") }
  }

  @Test
  fun `'addOrUpdateObject' throws with invalid parameters`() = runTest {
    val client = SearchClient(appId = "appId", apiKey = "apiKey")
    assertFails {
      client.addOrUpdateObject(
        indexName = empty(),
        objectID = "my-object-id",
        body = buildJsonObject {
        },
      )
    }.let { error -> assertError(error, "Parameter `indexName` is required when calling `addOrUpdateObject`.") }
    assertFails {
      client.addOrUpdateObject(
        indexName = "my-index-name",
        objectID = empty(),
        body = buildJsonObject {
        },
      )
    }.let { error -> assertError(error, "Parameter `objectID` is required when calling `addOrUpdateObject`.") }
    assertFails {
      client.addOrUpdateObject(
        indexName = "my-index-name",
        objectID = "my-object-id",
        body = empty(),
      )
    }.let { error -> assertError(error, "Parameter `body` is required when calling `addOrUpdateObject`.") }
  }
}
